oauthproxy.go

```diff
@@ -986,6 +999,11 @@ func (p *OAuthProxy) getAuthenticatedSession(rw http.ResponseWriter, req *http.R
 
 // addHeadersForProxying adds the appropriate headers the request / response for proxying
 func (p *OAuthProxy) addHeadersForProxying(rw http.ResponseWriter, req *http.Request, session *sessionsapi.SessionState) {
+	for name, values := range p.extraHeaders {
```

Comment:
I think we should make test cases for what happens when extraHeaders collide with headers other sections of the code would want to set.

Some things off the top of my head: Authorization, X-Forward-* user headers, X-Auth-Request-*, Cache-Control, Expires, X-Accel-Expires.

Do we want `extraHeaders` to take precedence over proxy generated ones? (potentially yes since people have been mentioning they want different cache settings for instance).

In which case this logic might need to be moved to later in the chain.

Reply:
@JoelSpeed Any thoughts on extra vs oauth2-proxy header collisions and precedence?

Reply:
Oh duh -- this only applies to request headers to the upstream.  I suppose no need to worry about broader collision with response headers.

I guess we just need to think about collisions with these set later in this method "Authorization, X-Forward-* user headers, X-Auth-Request-*".

I would think if a user sets one of those as an extraHeader - that still would take precedence though.  So maybe move this logic to the end of the method?

Reply:
Since we use `Add`, there shouldn't be a conflict, this adds the entry to the slice of values for a header if there already is, or sets the header to the value if there is. So if it clashes with something else we are setting, the final value of the header will be a comma separated list of the extra header value and the value set by us based on the claims in the session

Reply:
Which I actually don't think we want, right? IE what would backends do with a comma-separated `Authorization: (Basic|Bearer)` header for instance?

Reply:
While I can't think of a particular use case for this off the top of my head, I think users would soon realise what was happening if they found multiple values in their headers. I'd argue that would be a misconfiguration and maybe we shouldn't be preventing it, in case some people do actually want that behaviour. It would surprise me if people were adding headers that clashed with our own while expecting that configuration to work

Reply:
`Authorization` is maybe a bad example. A more likely scenario I can potentially see users wanting to do is overriding 1 of the `X-Forwarded-*` values to a hardcoded value the backend might need while still preserving the other ones.

Only non-stupid reason I could see for this is multiple oauth2-proxies fronting an app, normal one preserving user details and passing the user to the backend for it to authN. Second oauth2-proxy adjacent to it maybe overriding `X-Forwarded-User` for "Anonymous" user functionality. So we couldn't just set the second oauth2-proxy's extra header to another non-conflicting value -- it would need to align with the `X-Forwarded-User` of the first for the backend to do consistent validation.

Either way -- I feel like the spliced comma version isn't what I would expect at all. I would think we give 1 or the other a precedence and either validate and error if `--pass-users-header`, `--pass-authorization-header`, `--pass-access-token` are the preferred headers. Or extraHeaders takes precedence and we give a startup warning message that those hardcoded values are shadowing what would be set via claims.

Reply:
> Only non-stupid reason I could see for this is multiple oauth2-proxies fronting an app, normal one preserving user details and passing the user to the backend for it to authN. Second oauth2-proxy adjacent to it maybe overriding X-Forwarded-User for "Anonymous" user functionality. So we couldn't just set the second oauth2-proxy's extra header to another non-conflicting value -- it would need to align with the X-Forwarded-User of the first for the backend to do consistent validation.

In this case, would the second proxy just not be setting the User header? Like you'd configure it not to set it?

I think we need to gather some real world use cases to be able to make a decision on this. My temptation is to leave it as is with the comma separated values until someone raises an issue to explain why this doesn't work for their scenario, that I think would give us a better understanding of some real use and maybe help us make a decision